src/Build/Evaluation/Context/EvaluationContext.cs

```diff
@@ -66,11 +74,11 @@ internal EvaluationContext ContextForNewProject()
                 case SharingPolicy.Shared:
                     return this;
                 case SharingPolicy.Isolated:
-                    // use the first isolated context
-                    var used = Interlocked.CompareExchange(ref _used, 1, 0);
-                    return used == 0
+                    // reuse the first isolated context if it has not seen an evaluation yet.
+                    var previousValueWasUsed = Interlocked.CompareExchange(ref _used, 1, 0);
+                    return previousValueWasUsed == 0
```

Comment:
Shouldn't this condition be flipped?

Reply:
The problem actually was that the newly spawned isolated instance forgot that its coming from a line that has been previously used. This lead to every other isolated context to get erroneously reused.

Intended behaviour
1st isolated instance, first evaluation. Reuse the instance because it has not seen any instance
1st isolated instance, second evaluation. Spawn and use 2nd isolated instance
2nd isolated instance. Spawn and use 3rd isolated instance

Buggy behaviour
1st isolated instance, first evaluation. Reuse the instance because it has not seen any instance
1st isolated instance, second evaluation. Spawn and use 2nd isolated instance. But this instance forgot that it needs respawning.
2nd isolated instance. Since it forgot that it needs respawning, it gets reused